### PR TITLE
Modularize javascript

### PR DESCRIPTION
**Pull request by HelgeKrueger**
This pull request splits the main.js file into several node modules and uses browserify to recreate main.js. With node.js installed main.js can be created using `npm run bundle` inside the `web` folder.

See discussion at https://discuss.graphhopper.com/t/modularize-javascript-code/306/16

Installation:

```bash
curl -o- https://raw.githubusercontent.com/creationix/nvm/v0.29.0/install.sh | bash
# close and reopen terminal now
nvm install 4.2.2
nvm use 4.2.2

# git clone https://github.com/graphhopper/graphhopper.git
cd graphhopper/web
# download required packages:
npm install
npm test
# create main.js
npm run bundle

cd ..
./graphhopper.sh web africa_canary-islands.pbf
firefox localhost:8989
```

If you wanted to add the underscore library to the project, then you would use `npm install --save underscore` command. This command automatically modifies package.json to include underscore and a version info. This means that the next person running nvm install would also download underscore.

```bash
# For development just use watchify:
npm run watch
# see the package.json where several other scripts like bundleDebug are defined
# bundle creates the main file
npm run bundle
# create main.js for debugging
npm run bundleDebug
npm run bundleUgly

# Forcing consistent code style with jshint. JsHint can be run with:
npm run jshint
```

### CODE REVIEW

**Comment by karussell** [comment]
Thanks again :) ! 

Do we still provide an out of the box compiled main.js? Or do you think that this is a good solution?

**Comment by HelgeKrueger** [comment]
My last commit looks odd. It was meant to include the precompiled minified main.js. Let me try to fix it.

**Comment by HelgeKrueger** [comment]
I added the jasmine test to travis-ci and added jshint, which is style checker for JavaScript. I also added various missing semicolons, that jshint found.

File: .travis.yml [hunk 1]
```diff
@@ -5,10 +5,12 @@ jdk:
   - oraclejdk8
 # do not install anything instead return true via unix command true
 install: true
-script: ./core/files/travis-build.sh
+script:
+  - ./core/files/travis-build.sh
+  - ./web/files/travis-build.sh
```
Comment (karussell) [inline .travis.yml]: Hmmh, as the web jar is also built in the `core/files` we should either avoid a separate `web/files` script or put the web jar building into the `web/files` I fear :)

File: web/package.json [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+  "main": "main.js",
+  "scripts": {
+    "bundle": "browserify src/main/webapp/js/main-template.js -o src/main/webapp/js/main.js",
+    "bundleDebug": "browserify src/main/webapp/js/main-template.js --debug -o src/main/webapp/js/main.js",
+    "bundleUgly": "browserify -g uglifyify src/main/webapp/js/main-template.js -o src/main/webapp/js/main.js",
+    "test": "JASMINE_CONFIG_PATH=src/test/webapp/spec/jasmine.json jasmine",
+    "lint": "jshint src/main/webapp/js/"
+  },
+  "author": "",
+  "license": "ISC",
```
Comment (karussell) [inline web/package.json]: We should use Apache License 2.0 here as for all other GH stuff.

**Comment by karussell** [comment]
Thanks again for this work and also for making the tests working even on travis :) ! Is npm installed per default there?

Is leaflet.loading.js removed? See the functionality: https://github.com/ebrelsford/Leaflet.loading

**Comment by HelgeKrueger** [comment]
npm is installed by default on travis-ci.

File: web/package.json [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+  "dependencies": {
+    "browserify": "^12.0.1",
+    "d3": "^3.5.9",
+    "jasmine": "^2.3.2",
+    "jquery": "^2.1.4",
+    "leaflet": "^0.7.7",
+    "leaflet-loading": "^0.1.15",
```
Comment (HelgeKrueger) [inline web/package.json]: Leaflet loading inclusion part 1.

File: web/src/main/webapp/js/main-template.js [hunk 4]
```diff
@@ -0,0 +1,591 @@
...
+    if (location.port === '') {
+        host = location.protocol + '//' + location.hostname;
+    } else {
+        host = location.protocol + '//' + location.hostname + ":" + location.port;
+    }
+}
+
+global.d3 = require('d3');
+var L = require('leaflet');
+require('leaflet-loading');
```
Comment (HelgeKrueger) [inline web/src/main/webapp/js/main-template.js]: leaflet-loading inclusion part 2.

I was not able to use this npm inclusion mechanism for the other leaflet / jquery plugins. I also don't feel that this is super important at the moment to get all packages via npm.

**Comment by karussell** [comment]
Thanks!

I'm still struggeling to understand how I would enable 'debug mode'. It is still possible to change something in the source code and hit F5 without a build command in-between? 

And also I've added a bit installation instructions in your initial comment of this PR ... would you mind to explain the different bundle commands (see TODO)?

**Comment by karussell** [comment]
It looks like watchify is the solution (?)

```bash
npm install -g watchify
watchify src/main/webapp/js/main-template.js -o src/main/webapp/js/main.js
```

Is this good or what about beefy?

**Comment by karussell** [comment]
Merged it here https://github.com/graphhopper/graphhopper/commit/38e1c32069d1f296bbcbf3b5a09f00a6389b764c

One open question for me: "Also should we rename main.js to bundle.js and main-template.js to index.js or main.js to be more align with the normal namings for npm projects?"

**Pull request closed without merging**